### DISCUSSION TITLE
Github SSH not working

### DISCUSSION THREAD

**Discussion by HarryWangATX** [community/community#23063] [New to GitHub]
<p>Hi all,</p>
<p>When I run <code>ssh -T git@github.com</code> in my terminal, it connects to github. However, when I am pushing to my repository, it says <code>git@ssh.github.com: Permission denied (publickey).</code> I have checked that my remote does exist. Can anyone tell me how to solve this? Thanks!</p>

**Comment by queenofcorgis** [comment]
<p>:wave:t3: Welcome <a class="mention" href="/harrywangatx">@HarryWangATX</a>!</p>
<p>To get started:</p>
<ul>
<li>Is the output from the <code>ssh -T</code> command using your correct username?</li>
</ul>

**Comment by queenofcorgis** [comment]
<p>Also check out some documentation <a href="https://docs.github.com/en/authentication/troubleshooting-ssh/error-permission-denied-publickey">here</a> that may help :smile:</p>

**Comment by HarryWangATX** [comment]
<p>When my ssh-agent is started it outputs my username. However, when my ssh-agent hasn’t been started it gives the same error as above: <code>git@ssh.github.com: Permission denied (publickey).</code></p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="HarryWangATX" data-post="1" data-topic="262294">
<div class="title">
<div class="quote-controls"></div>
 HarryWangATX:</div>
<blockquote>
<p>When I run <code>ssh -T git@github.com</code> in my terminal, it connects to github. However, when I am pushing to my repository, it says <code>git@ssh.github.com: Permission denied (publickey).</code></p>
</blockquote>
</aside>
<p>There are different hostnames, <code>github.com</code> and <code>ssh.github.com</code>. The latter is for <a href="https://docs.github.com/en/authentication/troubleshooting-ssh/using-ssh-over-the-https-port">SSH over the HTTPS port</a>, which sometimes helps getting around blocks.</p>
<p>Considering that you can connect directly to <code>github.com</code> most likely you should fix the remote URL for your repository. If that doesn’t help, compare <code>ssh -v</code> output for both hosts.</p>

**Comment by HarryWangATX** [comment]
<p>My existing remote is pointed to the SSH URL. When I change it to the HTTPS URL it gives me this error when I push: <code>remote: Support for password authentication was removed on August 13, 2021. Please use a personal access token instead. remote: Please see https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/ for more information.</code></p>
<p>And it connected both to <code>git@github.com</code> and <code>git@ssh.github.com</code> successfully using <code>ssh -vT</code> on my terminal.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="HarryWangATX" data-post="6" data-topic="262294">
<div class="title">
<div class="quote-controls"></div>
 HarryWangATX:</div>
<blockquote>
<p>My existing remote is pointed to the SSH URL. When I change it to the HTTPS URL</p>
</blockquote>
</aside>
<p>I don’t mean to change it to HTTPS, I mean change it so it uses <code>github.com</code>, not <code>ssh.github.com</code>.</p>
<aside class="quote no-group" data-username="HarryWangATX" data-post="6" data-topic="262294">
<div class="title">
<div class="quote-controls"></div>
 HarryWangATX:</div>
<blockquote>
<p>And it connected both to <code>git@github.com</code> and <code>git@ssh.github.com</code> successfully using <code>ssh -vT</code> on my terminal.</p>
</blockquote>
</aside>
<p>That’s key information, in that case changing the host might not be enough. If it works when calling <code>ssh</code> directly but not through Git the only thing I can think of is to call Git with <a href="https://git-scm.com/docs/git#Documentation/git.txt-codeGITSSHCOMMANDcode"><code>GIT_SSH_COMMAND="ssh -v"</code></a> to get debug output from SSH as Git calls it.</p>
<p>Compare with the SSH tests you did, see what’s different (especially regarding used keys and authentication in general).</p>

**Comment by HarryWangATX** [comment, marked as answer]
<p>I tinkered with this for a bit just now, and I added a config file that made sure it wasn’t connecting to <a href="http://ssh.github.com" rel="noopener nofollow ugc">ssh.github.com</a> and instead <a href="http://github.com" rel="noopener nofollow ugc">github.com</a> as well as added the identity file for my public key and it works now! Thanks for the help!</p>

**Comment by HarryWangATX** [comment]
<aside class="quote group-Advisory-Council" data-username="airtower-luna" data-post="7" data-topic="262294">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p><a href="https://git-scm.com/docs/git#Documentation/git.txt-codeGITSSHCOMMANDcode" rel="noopener nofollow ugc"><code>GIT_SSH_COMMAND="ssh -v"</code></a></p>
</blockquote>
</aside>
<p>This is my output with ssh -v, my remote URL is already to <a href="mailto:git@github.com">git@github.com</a>:</p>
<pre><code class="lang-auto">❯ git push
OpenSSH_9.0p1, OpenSSL 1.1.1q  5 Jul 2022
debug1: Reading configuration data /c/Users/harry/.ssh/config
debug1: /c/Users/harry/.ssh/config line 1: Applying options for github.com
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Connecting to ssh.github.com [140.82.113.35] port 443.
debug1: Connection established.
debug1: identity file /c/Users/harry/.ssh/id_rsa type 0
debug1: identity file /c/Users/harry/.ssh/id_rsa-cert type -1
debug1: identity file /c/Users/harry/.ssh/id_ecdsa type -1
debug1: identity file /c/Users/harry/.ssh/id_ecdsa-cert type -1
debug1: identity file /c/Users/harry/.ssh/id_ecdsa_sk type -1
debug1: identity file /c/Users/harry/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /c/Users/harry/.ssh/id_ed25519 type 3
debug1: identity file /c/Users/harry/.ssh/id_ed25519-cert type -1
debug1: identity file /c/Users/harry/.ssh/id_ed25519_sk type -1
debug1: identity file /c/Users/harry/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /c/Users/harry/.ssh/id_xmss type -1
debug1: identity file /c/Users/harry/.ssh/id_xmss-cert type -1
debug1: identity file /c/Users/harry/.ssh/id_dsa type -1
debug1: identity file /c/Users/harry/.ssh/id_dsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_9.0
debug1: Remote protocol version 2.0, remote software version babeld-fc5e4bc0
debug1: compat_banner: no match: babeld-fc5e4bc0
debug1: Authenticating to ssh.github.com:443 as 'git'
debug1: load_hostkeys: fopen /c/Users/harry/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server-&gt;client cipher: chacha20-poly1305@openssh.com MAC: &lt;implicit&gt; compression: none
debug1: kex: client-&gt;server cipher: chacha20-poly1305@openssh.com MAC: &lt;implicit&gt; compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: SSH2_MSG_KEX_ECDH_REPLY received
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM
debug1: load_hostkeys: fopen /c/Users/harry/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: Host '[ssh.github.com]:443' is known and matches the ECDSA host key.
debug1: Found key in /c/Users/harry/.ssh/known_hosts:5
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: /c/Users/harry/.ssh/id_rsa RSA SHA256:yoGw6emIDNLcFKF+en4zUolSNhcmtIKmyYikDwXN2Hw
debug1: Will attempt key: /c/Users/harry/.ssh/id_ecdsa
debug1: Will attempt key: /c/Users/harry/.ssh/id_ecdsa_sk
debug1: Will attempt key: /c/Users/harry/.ssh/id_ed25519 ED25519 SHA256:5fLbQq9uWUhlLPOxr2FSr6IdWDyX/ScH5IqCm9UfxjE
debug1: Will attempt key: /c/Users/harry/.ssh/id_ed25519_sk
debug1: Will attempt key: /c/Users/harry/.ssh/id_xmss
debug1: Will attempt key: /c/Users/harry/.ssh/id_dsa
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=&lt;ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa&gt;
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey
debug1: Next authentication method: publickey
debug1: Offering public key: /c/Users/harry/.ssh/id_rsa RSA SHA256:yoGw6emIDNLcFKF+en4zUolSNhcmtIKmyYikDwXN2Hw
debug1: Authentications that can continue: publickey
debug1: Trying private key: /c/Users/harry/.ssh/id_ecdsa
debug1: Trying private key: /c/Users/harry/.ssh/id_ecdsa_sk
debug1: Offering public key: /c/Users/harry/.ssh/id_ed25519 ED25519 SHA256:5fLbQq9uWUhlLPOxr2FSr6IdWDyX/ScH5IqCm9UfxjE
debug1: Authentications that can continue: publickey
debug1: Trying private key: /c/Users/harry/.ssh/id_ed25519_sk
debug1: Trying private key: /c/Users/harry/.ssh/id_xmss
debug1: Trying private key: /c/Users/harry/.ssh/id_dsa
debug1: No more authentication methods to try.
git@ssh.github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
</code></pre>
